### PR TITLE
Feature/5124 implement additional service section view put endpoint 8147

### PR DESCRIPTION
**Pull request by SimonLLawrence**
Story [AB#5124](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/5124)
Task [AB#8147](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/8147)

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+using NHSD.BuyingCatalogue.Ordering.Api.Models;
+using NHSD.BuyingCatalogue.Ordering.Application.Persistence;
+using NHSD.BuyingCatalogue.Ordering.Api.Models.Summary;
+using NHSD.BuyingCatalogue.Ordering.Api.Services.CreateOrder;
+using NHSD.BuyingCatalogue.Ordering.Common.Constants;
+using NHSD.BuyingCatalogue.Ordering.Domain;
+
+namespace NHSD.BuyingCatalogue.Ordering.Api.Controllers
+{
+    [Route("api/v1/[controller]")]
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs:17]: I'm curious why you have added this when the canned data was in the OrderItemsController?

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            _orderRepository = orderRepository ?? throw new ArgumentNullException(nameof(orderRepository));
+        }
+
+        [HttpPut]
+        [Route("/api/v1/orders/{orderId}/sections/{sectionId}")]
+        [Authorize(Policy = PolicyName.CanManageOrders)]
+        public async Task<ActionResult> UpdateStatusAsync(string orderId, string sectionId, SectionStatusRequestModel sectionStatus )
+        {
+            if (sectionStatus == null)
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs:42]: is null

File: src/NHSD.BuyingCatalogue.Ordering.Api/Extensions/OrderExtensions.cs [hunk 3]
```diff
@@ -36,5 +36,14 @@ public static bool IsCatalogueSolutionsSectionComplete(this Order order)
             }
             return order.CatalogueSolutionsViewed;
         }
+
+        public static bool IsAdditionalServicesSectionComplete(this Order order)
+        {
+            if (order == null)
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Extensions/OrderExtensions.cs:42]: is null

File: src/NHSD.BuyingCatalogue.Ordering.Api/Extensions/OrderExtensions.cs [hunk 4]
```diff
@@ -36,5 +36,14 @@ public static bool IsCatalogueSolutionsSectionComplete(this Order order)
...
+            {
+                return false;
+            }
+            return order.AdditionalServicesViewed;
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Extensions/OrderExtensions.cs:46]: can you add a new line after the parenthesis

File: src/NHSD.BuyingCatalogue.Ordering.Api/Models/Summary/SectionStatusRequestModel.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace NHSD.BuyingCatalogue.Ordering.Api.Models.Summary
+{
+    public class SectionStatusRequestModel
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Models/Summary/SectionStatusRequestModel.cs:8]: sealed

File: src/NHSD.BuyingCatalogue.Ordering.Api/Models/Summary/SectionStatusRequestModel.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    {
+        public string Status { get; set; }
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Models/Summary/SectionStatusRequestModel.cs:10]: Also this class didnt need to be added as it was created in the canned data: 'UpdateOrderSectionModel"

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        public SectionStatusController(
+            IOrderRepository orderRepository)
+        {
+            _orderRepository = orderRepository ?? throw new ArgumentNullException(nameof(orderRepository));
+        }
+
+        [HttpPut]
+        [Route("/api/v1/orders/{orderId}/sections/{sectionId}")]
+        [Authorize(Policy = PolicyName.CanManageOrders)]
+        public async Task<ActionResult> UpdateStatusAsync(string orderId, string sectionId, SectionStatusRequestModel sectionStatus )
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs:40]: Can be linked to the UpdateOrderSectionModel instead

File: tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Order Summary.feature [hunk 8]
```diff
@@ -198,33 +198,56 @@ Scenario: 8. Get the order summary that includes a list of Catalogue Solutions
         | service-recipients  | incomplete | 2     |
         | catalogue-solutions | complete   | 2     |
         | additional-services | incomplete |       |
+        | funding-source      | incomplete |       |
+
+@7412
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Order Summary.feature:203]: Not sure what this number is referring too? Do you mean 5124?

File: tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Section Status Update.feature [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    As an Buyer User
+    I want to be able to update the section status for a given order section
+    So that I can ensure that the status is complete
+
+Background:
+    Given Orders exist
+        | OrderId    | Description         | OrderStatusId | AdditionalServicesViewed | CatalogueSolutionsViewed | Created    | LastUpdated | LastUpdatedBy                        | OrganisationId                       |
+        | C000014-01 | Some Description    | 1             | False                    | False                    | 11/05/2020 | 11/05/2020  | 335392e4-4bb1-413b-9de5-36a85c9c0422 | 4af62b99-638c-4247-875e-965239cd0c48 |
+    And the user is logged in with the Buyer role for organisation 4af62b99-638c-4247-875e-965239cd0c48
+
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Section Status Update.feature:11]: Remove a line

File: tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Section Status Update.feature [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+@5124
+Scenario Outline: 1. Set the additional services order section to completed
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Section Status Update.feature:14]: Just Scenario. Not Scenario outline

File: tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Section Status Update.feature [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    Then a response with status code 204 is returned
+    And the order with ID C000014-01 has additional services viewed set to true
+    And the order with ID C000014-01 has catalogue solutions viewed set to false
+    And the order with orderId C000014-01 is updated in the database with data
+        | LastUpdatedBy                        | LastUpdatedByName |
+        | 7b195137-6a59-4854-b118-62b39a3101ef | Bob Smith         |
+    And the order with orderId C000014-01 has LastUpdated time present and it is the current time
+
+@5124
+Scenario: 2. Set the catalog solutions order section to completed
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Features/Section Status Update.feature:25]: catalogue needs to be spelt correctly

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/SectionStatusControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            Assert.Throws<ArgumentNullException>(() => sectonControllerBuilder.Build());
+        }
+
+        [Test]
+        public async Task UpdateStatusAsync_OrderDoesNotExist_ReturnsNotFound()
+        {
+            var context = SectionStatusControllerTestContext.Setup();
+
+            var response =
+                await context.SectionStatusController.UpdateStatusAsync("INVALID", SectionModel.AdditionalServices.Id,new SectionStatusRequestModel{Status = "complete"});
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/SectionStatusControllerTests.cs:43]: can you add a space between the commas

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/SectionStatusControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        public async Task UpdateStatusAsync_WithAdditionalServicesSectionId_UpdatesAdditionalServicesViewedField()
+        {
+            var context = SectionStatusControllerTestContext.Setup();
+
+            const string orderId = "C0000014-01";
+            context.Order = CreateGetTestData(orderId, context.PrimaryOrganisationId, "ods");
+
+            var response = await context.SectionStatusController.UpdateStatusAsync(orderId, "additional-services", new SectionStatusRequestModel { Status = "complete" });
+            response.Should().BeOfType<NoContentResult>();
+            context.OrderRepositoryMock.Verify(u=>u.UpdateOrderAsync(It.Is<Order>(o=>o.AdditionalServicesViewed==true)));
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/SectionStatusControllerTests.cs:94]: space

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/SectionStatusControllerTests.cs [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+            {
+                (odsCode, "EU test")
+            };
+
+            repositoryOrder.AddOrderItem(repositoryOrderItem, Guid.Empty, string.Empty);
+            repositoryOrder.SetServiceRecipient(serviceRecipients, Guid.Empty, string.Empty);
+
+            return repositoryOrder;
+        }
+
```
Comment (RuthBerry96) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/SectionStatusControllerTests.cs:135]: Remove redundant spaces

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs (hunk 1 above)
Comment (SimonLLawrence) [reply src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs:17]: Following Sam advice on that one ha said it should be in separate controller.

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        };
+
+        public SectionStatusController(
+            IOrderRepository orderRepository)
+        {
+            _orderRepository = orderRepository ?? throw new ArgumentNullException(nameof(orderRepository));
+        }
+
+        [HttpPut]
+        [Route("/api/v1/orders/{orderId}/sections/{sectionId}")]
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs:38]: Sorry just noticed this. You've got a route at the top of the controller file and its being overwritten here. You only really need it in one place. Or then in the top have api/v2/orders
Then under the [HttpPut] include the rest

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs [hunk 16]
```diff
@@ -14,7 +14,7 @@
 
 namespace NHSD.BuyingCatalogue.Ordering.Api.Controllers
 {
-    [Route("api/v1/[controller]")]
+    [Route("/api/v1/orders")]
```
Comment (RuthBerry96) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/SectionStatusController.cs:17]: you don't need the / before api

**Pull request merged**